### PR TITLE
Primary key support for Arrow's `Memtable`

### PR DESCRIPTION
**Pull request by Jeadie**
# 🗣 Description
- For Arrow acceleration tables, enable support for handling primary keys for replace operations. 

## Required Documentation
 - https://github.com/spiceai/docs/pull/672

### CODE REVIEW

File: crates/data_components/src/arrow/write.rs [hunk 1]
```diff
@@ -17,19 +17,22 @@
 
 //! [`MemTable`] for querying `Vec<RecordBatch>` by `DataFusion`.
 
+use arrow::array::BooleanBuilder;
+use arrow::compute::filter_record_batch;
 use datafusion::catalog::Session;
 // This is modified from the DataFusion `MemTable` to support overwrites. This file can be removed once that change is upstreamed.
```
Comment (phillipleblanc) [inline crates/data_components/src/arrow/write.rs]:
```suggestion
```

I'm guessing we're probably not going to upstream it at this point.

File: crates/data_components/src/arrow/write.rs [hunk 2]
```diff
@@ -177,7 +180,19 @@ impl TableProvider for MemTable {
             ));
         }
 
-        let sink = Arc::new(MemSink::new(self.batches.clone(), overwrite));
+        // If the table has a primary key, we need to know it to handle `InsertOp::Replace`.
+        let mut primary_key = None;
+        if let Some(constraints) = self.constraints() {
```
Comment (phillipleblanc) [inline crates/data_components/src/arrow/write.rs]: This should return an error on constraints we don't handle, i.e. unique

File: crates/data_components/src/arrow/write.rs [hunk 3]
```diff
@@ -218,9 +235,84 @@ impl DisplayAs for MemSink {
...
+        Self {
+            batches,
+            overwrite,
+            primary_key,
+        }
+    }
+}
+
+/// Create primary key values for a [`RecordBatch`]. For composite keys, values are concatenated with a delimiter '|'.
+fn extract_primary_keys_str(batch: &RecordBatch, pk_indices: &[usize]) -> Result<Vec<String>> {
```
Comment (phillipleblanc) [inline crates/data_components/src/arrow/write.rs]: The order matters for the `pk_indices` - should we sort them to ensure they are always correct?

File: crates/data_components/src/arrow/write.rs
Comment (phillipleblanc) [inline crates/data_components/src/arrow/write.rs:1]: This file should have some tests for the new constraint functionality

File: crates/data_components/src/arrow/write.rs (hunk 1 above)
Comment (Jeadie) [reply crates/data_components/src/arrow/write.rs]: What's stopping us (besides getting interest about our changes?)

Comment (phillipleblanc) [reply crates/data_components/src/arrow/write.rs]: If we don't remove the comment then we can modify to include the changes we've made.

File: crates/data_components/src/arrow/write.rs [hunk 4]
```diff
@@ -218,9 +235,84 @@ impl DisplayAs for MemSink {
...
+
+    Ok(keys)
+}
+
+/// Filter elements of `existing_batches` that have primary keys from `overwriting_primary_keys`.
+///
+/// This is one part of `InsertOp::Replace` functionality, and still requires the new rows (with conflicting PKs), to be added.
+///
+/// This function modifies `existing_batches` in place.
+fn filter_existing(
```
Comment (Sevenannn) [inline crates/data_components/src/arrow/write.rs:293]: This function could use a test

File: crates/data_components/src/arrow/write.rs [hunk 5]
```diff
@@ -256,14 +371,68 @@ impl DataSink for MemSink {
...
+            new_key_set = new_primary_key_ids.into_iter().collect();
+            if new_key_set.len() != num_new_keys {
+                return Err(DataFusionError::Execution(
+                    "Primary key values must be unique".to_string(),
+                ));
+            }
+        }
+
+        // let mut writable_targets: Vec<_> =
+        //     futures::future::join_all(self.batches.iter().map(|target| target.write())).await;
```
Comment (phillipleblanc) [inline crates/data_components/src/arrow/write.rs]:
```suggestion
```

File: crates/data_components/src/arrow/write.rs [hunk 6]
```diff
@@ -256,14 +371,68 @@ impl DataSink for MemSink {
...
+                InsertOp::Replace => {
+                    if let Some(ref pks) = self.primary_key {
+                        filter_existing(&mut *target, &new_key_set, pks)?;
+                    }
+                }
             }
-            target.append(&mut batches);
+
+            // println!("New batch {}", pretty_format_batches(&batches).unwrap());
+
```
Comment (phillipleblanc) [inline crates/data_components/src/arrow/write.rs]:
```suggestion
```

File: crates/data_components/src/arrow/write.rs [hunk 7]
```diff
@@ -256,14 +371,68 @@ impl DataSink for MemSink {
...
+            target.extend(batches.drain(..));
+            // println!("target after {}", pretty_format_batches(target).unwrap());
+            // println!(
+            //     "New batch after {}",
+            //     pretty_format_batches(&batches).unwrap()
+            // );
```
Comment (phillipleblanc) [inline crates/data_components/src/arrow/write.rs]:
```suggestion
```

File: crates/data_components/src/arrow/write.rs [hunk 8]
```diff
@@ -256,14 +371,68 @@ impl DataSink for MemSink {
...
+        }
+
+        for b in &self.batches {
+            let part = b.read().await;
+            println!("self.batches \n{}\n", pretty_format_batches(&part).unwrap());
```
Comment (phillipleblanc) [inline crates/data_components/src/arrow/write.rs]:
```suggestion
```

**Review by phillipleblanc — DISMISSED** [review]

**Review by phillipleblanc — DISMISSED** [review]

**Review by sgrebnov — DISMISSED** [review]

**Review by Sevenannn — DISMISSED** [review]

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**